### PR TITLE
Nightcycle manager new control types and zone lists

### PR DESCRIPTION
**Pull request by mjwitte**
## Pull request overview

Adds new control type options and zone lists to AvailabilityManager:NightCycle

New control types: 
CycleOnAnyCoolingOrHeatingZone, CycleOnAnyCoolingZone, CycleOnAnyHeatingZone, and CycleOnAnyHeatingZoneFansOnly 

New zone list fields:
Control Zone or Zone List Name _(formerly Control Zone Name)_
Cooling Control Zone or Zone List Name
Heating Control Zone or Zone List Name
Heating Zone Fans Only Zone or Zone List Name

No transition necessary.
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [ ] If defect, results of running current develop vs this branch should exhibit the fix
- [ ] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [ ] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] Open windows IDF Editor with modified IDD to check for errors
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] Changed docs build successfully
- [ ] ExpandObjects changes?
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by mjwitte** [comment]
@Myoldmopar (or @EnergyArchmage ) this is ready for review.

**Comment by mjwitte** [comment]
The diffs in PIUAuto are expected, because it was altered significantly for the purpose of testing the new night cycle zone lists.

**Comment by Myoldmopar** [comment]
This didn't result in any conflicts after ASIHP went in, so after CI gets a start on that merge, this will go in right after.

File: doc/engineering-reference/src/simulation-models-encyclopedic-reference-004/system-availability-managers.tex [hunk 1]
```diff
@@ -30,31 +30,55 @@ \subsection{Night Cycle}\label{night-cycle}
 
 The input object AvailabilityManager:NightCycle is to specify when a system that has been scheduled off during unoccupied hours should cycle on to prevent building temperatures from becoming too hot or too cold. This manager can not be used to force a system off. The inputs are:
 
-·~~~~~~~~A manager applicability schedule;
+\begin{itemize}
```
Comment (Myoldmopar) [inline doc/engineering-reference/src/simulation-models-encyclopedic-reference-004/system-availability-managers.tex:33]: 💯 for cleanup!

File: doc/input-output-reference/src/overview/group-system-availability-managers.tex [hunk 2]
```diff
@@ -143,7 +145,19 @@ \subsection{AvailabilityManager:NightCycle}\label{availabilitymanagernightcycle}
 
 \textbf{\emph{Field: Control Type}}
 
-The possible inputs are \emph{StayOff}, \emph{CycleOnAny}, \emph{CycleOnControlZone}, \emph{CycleOnAnyZoneFansOnly}. \emph{StayOff} means the availability manager will have no effect air handler on/off will be determined by the fan schedule. \emph{CycleOnAny} means that if any zone served by a system whose AvailabilityManagerAssignmentList references this system availability manager has an air temperature outside the cooling or heating set points the central fan will turn on even though the fan schedule indicates the fan is off. \emph{CycleOnControlZone} means the same thing except the availability manager looks at the temperature in only one zone. \emph{CycleOnAnyZoneFansOnly} is the same as \emph{CycleOnAny} except that only the zone fans are cycled on and the central air handler fan is left off. The default is \emph{StayOff}.
+The possible inputs are \emph{StayOff}, \emph{CycleOnAny}, \emph{CycleOnControlZone}, \emph{CycleOnAnyZoneFansOnly}, \emph{CycleOnAnyCoolingOrHeatingZone}, \emph{CycleOnAnyCoolingZone}, \emph{CycleOnAnyHeatingZone}, or \emph{CycleOnAnyHeatingZoneFansOnly}. The default is \emph{StayOff}.
+
+\begin{itemize}
```
Comment (Myoldmopar) [inline doc/input-output-reference/src/overview/group-system-availability-managers.tex:150]: Just a note, in addition to the `itemize` environment, there is also the [`description`](https://en.wikibooks.org/wiki/LaTeX/List_Structures#List_structures) environment that can do this sort of "keyword followed by description" structure.  Definitely no need to change anything, just noting it for future doc updates.

File: doc/engineering-reference/src/simulation-models-encyclopedic-reference-004/system-availability-managers.tex (hunk 1 above)
Comment (mjwitte) [inline doc/engineering-reference/src/simulation-models-encyclopedic-reference-004/system-availability-managers.tex:33]: Even this dog can learn some new LaTeX tricks!  And did you notice that [Building Simulation 2017](http://www.bs2017.org/instructions.html) accepts abstracts in LaTeX? Scroll down and there's a LaTeX abstract template.

Comment (Myoldmopar) [inline doc/engineering-reference/src/simulation-models-encyclopedic-reference-004/system-availability-managers.tex:33]: 😍

File: src/EnergyPlus/SystemAvailabilityManager.cc [hunk 3]
```diff
@@ -2201,60 +2305,104 @@ namespace SystemAvailabilityManager {
...
+	CoolingZoneOutOfTolerance(
+		Array1D_int const ZonePtrList, // list of controlled zone pointers
+		int const NumZones, // number of zones in list
+		Real64 const TempTolerance // temperature tolerance
+	)
+	{
+		// Check if any zone temperature is above the cooling setpoint plus tolerance
+		for (int Index = 1; Index <= NumZones; ++Index) { // loop over zones in list
+			int ZoneNum = ZonePtrList( Index );
+			{ auto const tstatType ( DataHeatBalFanSys::TempControlType( ZoneNum ) );
```
Comment (Myoldmopar) [inline src/EnergyPlus/SystemAvailabilityManager.cc:2368]: Speaking of new tricks, for future code keep in mind the phrase: "class enums and switch blocks."

Comment (mjwitte) [inline src/EnergyPlus/SystemAvailabilityManager.cc:2368]: Yeah, thought about enums.  That can happen on a second pass.  It's ripe for some other restructuring, too.

Comment (Myoldmopar) [inline src/EnergyPlus/SystemAvailabilityManager.cc:2368]: Definitely; just making notes for later.

**Comment by Myoldmopar** [comment]
Pulled this and built and tested locally.  Everything looks good.  Code looks good.  CI has started working on the last develop commit, so I'm dropping this now.

**Pull request merged**